### ISSUE TITLE
Ingress pathType: Prefix returning 404

### ISSUE THREAD

**Issue by micya** [openservicemesh/osm#3292]
**Bug description**:

Creating Kubernetes ingress resource with pathType: Prefix, additional path parts don't route properly and returns 404 incorrectly.

For example, when ingress path: /, /get incorrectly returns 404 (172.18.0.2 is nginx ingress controller LoadBalancer IP):

```bash
$ curl -v 172.18.0.2/get
*   Trying 172.18.0.2...
* TCP_NODELAY set
* Connected to 172.18.0.2 (172.18.0.2) port 80 (#0)
> GET /get HTTP/1.1
> Host: 172.18.0.2
> User-Agent: curl/7.58.0
> Accept: */*
> 
< HTTP/1.1 404 Not Found
< Date: Tue, 04 May 2021 20:29:01 GMT
< Content-Length: 0
< Connection: keep-alive
< 
* Connection #0 to host 172.18.0.2 left intact
```

**Affected area (please mark with X where applicable)**:
- Install                    [ ]
- SMI Traffic Access Policy  [ ]
- SMI Traffic Specs Policy   [ ]
- SMI Traffic Split Policy   [ ]
- Permissive Traffic Policy  [ ]
- Ingress                    [ X ]
- Egress                     [ ]
- Envoy Control Plane        [ ]
- CLI Tool                   [ ]
- Metrics                    [ ]
- Certificate Management     [ ]
- Sidecar Injection          [ ]
- Logging                    [ ]
- Debugging                  [ ]
- Tests                      [ ]
- Demo                       [ ]
- CI System                  [ ]

**Expected behavior**:

```bash
$ curl -v 172.18.0.2/get
*   Trying 172.18.0.2...
* TCP_NODELAY set
* Connected to 172.18.0.2 (172.18.0.2) port 80 (#0)
> GET /get HTTP/1.1
> Host: 172.18.0.2
> User-Agent: curl/7.58.0
> Accept: */*
> 
< HTTP/1.1 200 OK
< Date: Tue, 04 May 2021 20:29:16 GMT
< Content-Type: application/json
< Content-Length: 259
< Connection: keep-alive
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Credentials: true
< 
{
  "args": {}, 
  "headers": {
    "Accept": "*/*", 
    "Host": "172.18.0.2", 
    "User-Agent": "curl/7.58.0", 
    "X-Forwarded-Host": "172.18.0.2", 
    "X-Scheme": "http"
  }, 
  "origin": "10.42.0.1", 
  "url": "http://172.18.0.2/get"
}
* Connection #0 to host 172.18.0.2 left intact
```

**Steps to reproduce the bug (as precisely as possible)**:

1. Install osm:

```bash
osm install
```

2. Install nginx ingress

```bash
kubectl create namespace ingress-basic
helm install nginx-ingress ingress-nginx/ingress-nginx \
    --namespace ingress-basic
```

3. Create deployment for httpbin:

```bash
# Create the httpbin namespace
kubectl create namespace httpbin

# Add the namespace to the mesh
osm namespace add httpbin

# Deploy httpbin service in the httpbin namespace
kubectl apply -n httpbin -f - << EOF
apiVersion: v1
kind: ServiceAccount
metadata:
  name: httpbin
---
apiVersion: v1
kind: Service
metadata:
  name: httpbin
  labels:
    app: httpbin
    service: httpbin
spec:
  ports:
  - name: http
    port: 14001
    targetPort: 80
  selector:
    app: httpbin
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: httpbin
spec:
  replicas: 1
  selector:
    matchLabels:
      app: httpbin
  template:
    metadata:
      labels:
        app: httpbin
    spec:
      serviceAccountName: httpbin
      containers:
      - image: kennethreitz/httpbin
        imagePullPolicy: IfNotPresent
        name: httpbin
        ports:
        - containerPort: 80
EOF
```

4. Create ingress:

```bash
kubectl apply -f - << EOF
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: httpbin-ingress
  namespace: httpbin
spec:
  ingressClassName: nginx
  rules:
  - http:
      paths:
      - path: /
        pathType: Prefix
        backend:
          service:
            name: httpbin
            port:
              number: 14001
EOF
```

5. Curl

This works correctly (172.18.0.2 is ingress controller IP):

```bash
curl 172.18.0.2
```

This does NOT work (but should work):

```bash
$ curl -v 172.18.0.2/get
*   Trying 172.18.0.2...
* TCP_NODELAY set
* Connected to 172.18.0.2 (172.18.0.2) port 80(#0)
> GET /get HTTP/1.1
> Host: 172.18.0.2
> User-Agent: curl/7.58.0
> Accept: */*
> 
< HTTP/1.1 404 Not Found
< Date: Tue, 04 May 2021 20:24:18 GMT
< Content-Length: 0
< Connection: keep-alive
< 
* Connection #0 to host 172.18.0.2 left intact
```

**How was OSM installed?**:

Using osm cli:

```bash
osm install
```

**Anything else we need to know?**:

Logs from envoy sidecar of httpbin pod:

```bash
{"bytes_received":0,"protocol":"HTTP/1.1","x_forwarded_for":"10.42.0.1","user_agent":"curl/7.58.0","requested_server_name":null,"start_time":"2021-05-04T20:15:38.772Z","duration":0,"upstream_cluster":null,"bytes_sent":0,"upstream_host":null,"authority":"172.18.0.2:80","time_to_first_byte":null,"method":"GET","request_id":"091398d96c15fff7d7ac6d98413de6ac","response_code":404,"response_flags":"NR","upstream_service_time":null,"response_code_details":"route_not_found","path":"/get"}
```

Ingress controller logs:

```
10.42.0.1 - - [04/May/2021:20:15:38 +0000] "GET /get HTTP/1.1" 404 0 "-" "curl/7.58.0" 82 0.002 [httpbin-httpbin-14001] [] 10.42.1.12:80 0 0.004 404 091398d96c15fff7d7ac6d98413de6ac
```

**Environment**:
- OSM version (use `osm version`):
  - Version: v0.8.3; Commit: 6fa4a5791cd70ec5eb2f0abaa0c09b7993c2fd34; Date: 2021-04-20-21:44
- Kubernetes version (use `kubectl version`): 
  - k3d version v4.4.0
  - k3s version v1.20.4-k3s1 (default)
- Size of cluster (number of worker nodes in the cluster):
  - 1 master, 2 worker
- Others:

**Comment by shashankram** [comment]
Thanks @micya for reporting this, I'll look into this.

**Comment by shashankram** [comment]
Hey @micya, there are 2 reasons why this won't work:
1. Ingress pathType matching is not supported in OSM v0.8.x, this feature is only available to use in the next release of OSM v0.9.0
2. There is a caveat with how Envoy performs regex matching on certain paths, such as the wildcard prefix `/`. https://github.com/openservicemesh/osm/pull/3296 should address this.

So to be able to use Ingress V1 and a wildcard path `/` whose `pathType=Prefix`, one needs to use v0.9.0 (release date is TBD).

**Comment by shashankram** [comment]
The codebase on `main` branch should have this fixed:

```
$ kubectl apply -f - << EOF
> apiVersion: networking.k8s.io/v1
> kind: Ingress
> metadata:
>   name: httpbin-ingress
>   namespace: httpbin
> spec:
>   ingressClassName: nginx
>   rules:
>   - http:
>       paths:
>       - path: /
>         pathType: Prefix
>         backend:
>           service:
>             name: httpbin
>             port:
>               number: 14001
> EOF
ingress.networking.k8s.io/httpbin-ingress configured
$ 
$ curl -I http://20.75.121.115/status/200
HTTP/1.1 200 OK
Date: Wed, 05 May 2021 01:07:09 GMT
Content-Type: text/html; charset=utf-8
Content-Length: 0
Connection: keep-alive
access-control-allow-origin: *
access-control-allow-credentials: true
x-envoy-upstream-service-time: 1

$ curl -I http://20.75.121.115/get
HTTP/1.1 200 OK
Date: Wed, 05 May 2021 01:07:14 GMT
Content-Type: application/json
Content-Length: 363
Connection: keep-alive
access-control-allow-origin: *
access-control-allow-credentials: true
x-envoy-upstream-service-time: 2
```
